### PR TITLE
Rework extract functions.

### PR DESCRIPTION
**Pull request by Joroks**
I've reworked my previous PR with a more focused scope. It's still quite large so I've tried my best to make it as polished as possible to make reviewing it easier / reduce the amount of revisions needed for it to be merged.

I do feel a bit sorry for comming out of nowhere and suddendly putting this huge amount of extra workload on your table 😅. I'm really gratefull for the work you've put into this thus far! Please take as much time as you need to review this.

### CODE REVIEW

**Comment by Joroks** [comment]
Of course there is an error on every operating system except for the one I'm running on ...

**Comment by vchuravy** [comment]
You might need to run under Valgrind.

File: src/LAMMPS.jl [hunk 1]
```diff
@@ -97,13 +154,11 @@ end
...
 Create a new LAMMPS instance and call `f` on that instance while returning the result from `f`.
-This constructor closes the LAMMPS instance immediately after `f` has executed.
 """
 function LMP(f::Function, args=String[], comm=nothing)
     lmp = LMP(args, comm)
-    result = f(lmp)
-    close!(lmp)
-    return result
+    return f(lmp)
+    # `close!` is registered as a finalizer for LMP, no need to close it here.
```
Comment (vchuravy) [inline src/LAMMPS.jl:161]: Can you split this out to its own commit? This should be quick to land

**Comment by Joroks** [comment]
@vchuravy the error is now fixed. It actually was an issue with the C-API documentation (https://docs.lammps.org/Library_objects.html#_CPPv423lammps_extract_variablePvPKcPKc). It names two strings to acces the length of the vector:

- "GET_VECTOR_SIZE"
- "LMP_SIZE_VECTOR"

And I happened to pick the wrong one.

if we pass "LMP_SIZE_VECTOR", we're expected to dereference it.
if we pass anything else we are explicitly told not to dereference it.

So that's were the error came from

File: src/LAMMPS.jl (hunk 1 above)
Comment (vchuravy) [reply src/LAMMPS.jl:161]: Yeah separate PRs

**Comment by vchuravy** [comment]
Can you open an upstream PR to fix the docs?

File: src/LAMMPS.jl [hunk 2]
```diff
@@ -1,12 +1,69 @@
...
+       SIZE_COLS,
+
+       VAR_EQUAL,
+       VAR_ATOM,
+       VAR_VECTOR,
+       VAR_STRING,
+
+       LMP_STYLE_GLOBAL,
+       LMP_STYLE_ATOM,
+       LMP_STYLE_LOCAL
```
Comment (vchuravy) [inline src/LAMMPS.jl:33]: I would almost consider not exporting them.

And then name then `NONE` instead of `LAMMPS_NONE` and `STYLE_GLOBAL` instead of `LMP_STYLE_GLOBAL`.
Then it's consitent in user code as `LAMMPS.NONE` and `LAMMPS.STYLE_GLOBAL`

File: src/LAMMPS.jl [hunk 3]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
-        shape = reverse(shape)
-    end
+    ndata == 0 && return Matrix{T}(undef, count, ndata) # There is no data that can be wrapped
+
+    pointers = Base.unsafe_wrap(Array, ptr, ndata)
+
+    # This assert verifies that all the pointers are evenly spaced according to count.
+    # While it seems like this is allways the case, it's not explicitly stated in the
+    # API documentation. It also helps to verify, that the count is correct.
+    @assert all(diff(pointers) .== count*sizeof(T))
```
Comment (vchuravy) [inline src/LAMMPS.jl:264]: This assert is quite costly. I would only do so in some sort of special debug mode.

```suggestion
    # @assert all(diff(pointers) .== count*sizeof(T))
```

File: src/LAMMPS.jl [hunk 4]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
-        @assert length(shape) == 2
+    (count, ndata) = shape
 
-        # Note: Julia like Fortran is column-major
-        #       so the data is transposed from Julia's perspective
-        shape = reverse(shape)
-    end
+    ndata == 0 && return Matrix{T}(undef, count, ndata) # There is no data that can be wrapped
+
+    pointers = Base.unsafe_wrap(Array, ptr, ndata)
```
Comment (vchuravy) [inline src/LAMMPS.jl:259]: Since we immediatly load from this ptr. We allocate an Array structure that we immediately discard.

File: src/LAMMPS.jl [hunk 5]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
-    end
+    return copy ? Base.copy(result) : result
 end
 
-function unsafe_wrap(ptr, shape)
-    if length(shape) > 1
-        # We got a list of ptrs,
-        # but the first pointer points to the whole data
-        ptr = Base.unsafe_load(ptr)
+function _lammps_wrap(ptr::Ptr{<:Ptr{T}}, shape::NTuple{2}, copy=true) where T
```
Comment (vchuravy) [inline src/LAMMPS.jl:252]: I do think this `wrap` function should not take a copy argument. Wrap implies to me that the memory is shared.

File: src/LAMMPS.jl [hunk 6]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
-    if axes2 !== nothing
-        shape = (axes1, axes2)
+A full list of global variables can be found here: <https://docs.lammps.org/Library_properties.html>
+"""
+function extract_global(lmp::LMP, name::String, lmp_type::_LMP_DATATYPE; copy::Bool=true)
+    void_ptr = API.lammps_extract_global(lmp, name)
+    void_ptr == C_NULL && error("Unknown global variable $name")
+
+    expect = extract_global_datatype(lmp, name)
+    recieve = get_enum(lmp_type)
```
Comment (vchuravy) [inline src/LAMMPS.jl:343]:
```suggestion
    receive = get_enum(lmp_type)
```

File: src/LAMMPS.jl [hunk 7]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
-        val = Base.unsafe_load(Base.unsafe_convert(Ptr{Float64}, ptr))
-        API.lammps_free(ptr)
-        return val
-    elseif var == API.LMP_VAR_ATOM
-        nlocal = extract_global(lmp, "nlocal")
-        ptr = API.lammps_extract_variable(lmp, name, group)
-        if ptr == C_NULL
-            error("Group $group for variable $name with style atom not available.")
+    void_ptr = API.lammps_extract_variable(lmp, name, group)
+    void_ptr == C_NULL && error("Unknown variable $name")
```
Comment (vchuravy) [inline src/LAMMPS.jl:525]: Any reason not to use a `KeyError`?

File: src/api.jl [hunk 8]
```diff
@@ -6,7 +6,8 @@ import LAMMPS_jll
 import LAMMPS_jll: liblammps
 import MPI: MPI_Comm
 
-@cenum _LMP_DATATYPE_CONST::UInt32 begin
+@cenum _LMP_DATATYPE_CONST::Int32 begin
```
Comment (vchuravy) [inline src/api.jl:9]: This file is auto generated, in https://github.com/cesmix-mit/LAMMPS.jl/tree/main/res

**Review by vchuravy — COMMENTED** [review]
I do think that `extract` should maybe have `copy=false`.

File: src/LAMMPS.jl [hunk 9]
```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
...
+        error("TypeMismatch: Expected $expect got $recieve instead!")
+    end
+
+    if lmp_variable == VAR_EQUAL
+        ptr = _lammps_reinterpret(LAMMPS_DOUBLE, void_ptr)
+        result = unsafe_load(ptr)
         API.lammps_free(ptr)
-        return val
-    elseif var == API.LMP_VAR_VECTOR
-        # TODO Fix lammps docs `GET_VECTOR_SIZE`
```
Comment (vchuravy) [inline src/LAMMPS.jl:401]: Ah I ran into the same issue before....

**Comment by vchuravy** [comment]
Also bump the `Project.toml` minor version.

File: src/api.jl (hunk 8 above)
Comment (Joroks) [reply src/api.jl:9]: For some reason the generated file doesn't fully match the lammps source code. From lammps/src/library.h:

```C
enum _LMP_DATATYPE_CONST {
  LAMMPS_NONE = -1,     /*!< no data type assigned (yet) */
  LAMMPS_INT = 0,       /*!< 32-bit integer (array) */
  LAMMPS_INT_2D = 1,    /*!< two-dimensional 32-bit integer array */
  LAMMPS_DOUBLE = 2,    /*!< 64-bit double (array) */
  LAMMPS_DOUBLE_2D = 3, /*!< two-dimensional 64-bit double array */
  LAMMPS_INT64 = 4,     /*!< 64-bit integer (array) */
  LAMMPS_INT64_2D = 5,  /*!< two-dimensional 64-bit integer array */
  LAMMPS_STRING = 6     /*!< C-String */
};
```

LAMMPS_NONE is missing in the generated file. Which is problematic since it breaks

```julia
function extract_global_datatype(lmp::LMP, name)
    return API._LMP_DATATYPE_CONST(API.lammps_extract_global_datatype(lmp, name))
end
```

as `API.lammps_extract_global_datatype` returns -1 if the name is not found.

Comment (vchuravy) [reply src/api.jl:9]: This likely means that the file was generated from an older version of that header. So we need to update it, if you can do that in a separate PR that would be great.

**Comment by vchuravy** [comment]
> the better solution is to do this

I think that's fine. Better than exporting the `STRING`

File: src/LAMMPS.jl [hunk 10]
```diff
@@ -177,240 +245,362 @@ function get_natoms(lmp::LMP)
...
+
+    # If the pointers are evenly spaced, we can simply use the first pointer to wrap our matrix.
+    first_pointer = unsafe_load(ptr)
+    return Base.unsafe_wrap(Array, first_pointer, shape, own=own)
+
+end
+
+function _lammps_copy(ptr::Ptr{T}, shape::Integer) where T <: Real
+    ptr == C_NULL && error("Copying NULL-pointer!")
+    return [unsafe_load(ptr, i) for i in 1:shape]
```
Comment (vchuravy) [inline src/LAMMPS.jl:275]: I don't think it's cleaner to do:

return `copy(_lammps_wrap())`

I know a bit counter to my comment earlier, but copy can lower to memcpy directly instead of hoping for the compiler to recognize and optimize this call.

File: src/LAMMPS.jl [hunk 11]
```diff
@@ -177,240 +245,362 @@ function get_natoms(lmp::LMP)
...
-    dtype = API._LMP_DATATYPE_CONST(dtype)
-    type = dtype2type(dtype)
+function _lammps_wrap(ptr::Ptr{<:Real}, shape::Integer; own=false)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    return Base.unsafe_wrap(Array, ptr, shape, own=own)
+end
 
-    ptr = API.lammps_extract_global(lmp, name)
-    ptr = reinterpret(type, ptr)
+function _lammps_wrap(ptr::Ptr{<:Ptr{T}}, shape::NTuple{2}; own=false) where T
```
Comment (vchuravy) [inline src/LAMMPS.jl:258]: I don't particularly like the naming< we are in the LAMMPS module, so adding LAMMPS feels superfluous.

`unsafe_wrap` or just wrap

**Review by vchuravy — APPROVED** [review]
Besides the two small nits, this looks great to me.

I would also like to invite you as a collaborator to this repository. The expectation would still be to use PRs and to wait/ask for review before merging large changes, but by now you rewritten most of the code here :)

File: src/LAMMPS.jl (hunk 11 above)
Comment (Joroks) [reply src/LAMMPS.jl:258]: I think It's best to just roll these two functions into one again but call it `_extract` instead of `wrap`.

we basically allways do
```julia
copy ? _lammps_copy(ptr, length) : _lammps_wrap(ptr, length)
```
afterall.

**Comment by Joroks** [comment]
> I would also like to invite you as a collaborator to this repository. The expectation would still be to use PRs and to wait/ask for review before merging large changes, but by now you rewritten most of the code here :)

I'll gladly accept this invitation!

File: src/LAMMPS.jl (hunk 11 above)
Comment (vchuravy) [reply src/LAMMPS.jl]: I don't think we need `_copy`

Is there a material benefit over. `copy(unsafe_wrap())`

I do expect advanced users being exposed to wrap, since they might use the LAMMPS API directly

File: src/LAMMPS.jl [hunk 12]
```diff
@@ -177,240 +245,344 @@ function get_natoms(lmp::LMP)
...
-function unsafe_wrap(ptr, shape)
-    if length(shape) > 1
-        # We got a list of ptrs,
-        # but the first pointer points to the whole data
-        ptr = Base.unsafe_load(ptr)
+function _extract(ptr::Ptr{<:Ptr{T}}, shape::NTuple{2}; copy=false, own=false) where T
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
 
-        @assert length(shape) == 2
+    shape[2] == 0 && return Matrix{T}(undef, shape) # There is no data that can be wrapped
```
Comment (vchuravy) [inline src/LAMMPS.jl:263]:
```suggestion
    prod(shape) == 0 && return Matrix{T}(undef, shape) # There is no data that can be wrapped
```

File: src/LAMMPS.jl [hunk 13]
```diff
@@ -177,240 +245,344 @@ function get_natoms(lmp::LMP)
...
-"""
-    extract_global(lmp, name, dtype=nothing)
-"""
-function extract_global(lmp::LMP, name, dtype=nothing)
-    if dtype === nothing
-        dtype = API.lammps_extract_global_datatype(lmp, name)
-    end
-    dtype = API._LMP_DATATYPE_CONST(dtype)
-    type = dtype2type(dtype)
+function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
```
Comment (vchuravy) [inline src/LAMMPS.jl:253]: `copy` should imply `own`?  We can almost never use `own=true` since we need to use lammps_free for the deallocation

Comment (Joroks) [reply src/LAMMPS.jl:253]: As far as I can tell, we don't need to use `lammps_free` for deallocation. It just calls `free` internally without doing anything else. Thats the same thing that would happen If we take ownership of the memory if I'm not mistaken. I've taken this directly from the lammps source code:
```C++
void lammps_free(void *ptr)
{
  free(ptr);
}
```

The `own` keyword is used for the `extract_variable` method, specifically when it's a Atom style variable:
```julia
  if lmp_variable == VAR_ATOM
      ndata = extract_setting(lmp, "nlocal")
      ptr = _reinterpret(LAMMPS_DOUBLE, void_ptr)
      # lammps expects us to take ownership of the data
      return _extract(ptr, ndata; own=true)
  end
```
because the documentation states:
> For atom-style variables, the return value is a pointer to an allocated block of storage of double of the length
atom->nlocal. Since the data returned are a copy, the location will persist, but its content will not be updated in case the variable is re-evaluated. To avoid a memory leak, this pointer needs to be freed after use in the calling program.

That means that the value is already a copy and copying it again would be superfluous.

I've also (crudely) verified, that this does indeed not cause a memory leak by running the following script:

```julia
lmp = LMP()

command(lmp, """
         atom_modify map yes
         region cell block 0 3 0 3 0 3
         create_box 1 cell
         lattice sc 1
         create_atoms 1 region cell
         mass 1 1
        
         variable var atom id
    """)

for _ in 1:10^7
    extract_variable(lmp, "var", VAR_ATOM)
end
```

If I set `own` to false, my memory usage rises by about 2GB. If I set `own` to true the memory consumption stays stable.
So I'd say its safe to let the GC handle freeing the memory if we're supposed to take ownership.

I have also verified that the data stays allocated even if we close the lammps instance. I've even written this as a test:

```julia
lmp = LMP(["-screen", "none"])
command(lmp, """
    atom_modify map yes
    region cell block 0 3 0 3 0 3
    create_box 1 cell
    lattice sc 1
    create_atoms 1 region cell
    mass 1 1

    variable var atom id
""")

var = extract_variable(lmp, "var", VAR_ATOM)
var_copy = copy(var)
LAMMPS.close!(lmp)

@test var == var_copy
```

In summary, we use `copy=true`, if the ownership stays with LAMMPS but we want a copy and `own=true` if the ownership is handed to us, which is usually because LAMMPS has made a copy for us

Comment (vchuravy) [reply src/LAMMPS.jl:253]: > t just calls free internally without doing anything else. Thats the same thing that would happen If we take ownership of the memory if I'm not mistaken.

This is sadly not true. In shared libraries the `malloc` and `free` maybe be resolved to different implementations. So you can never rely that the functions are the same in two shared libraries.

> if we're supposed to take ownership.

So that's actually a good question. When are we supposed to take ownership? 

As far as I understood `extract` we should never take ownershi, but if the underlying API is allocating memory for us, then we will need to do a `copy` + `lammps_free`. Which sucks.

Comment (vchuravy) [reply src/LAMMPS.jl:253]: Okay according to the docs https://docs.lammps.org/Library_objects.html
atom style requires us to call `lammps_free`.

Comment (Joroks) [reply src/LAMMPS.jl:253]: I've thought about registering lammps_free as a finalizer on our array. I think that could work?

Comment (Joroks) [reply src/LAMMPS.jl:253]: Doing this

```julia
function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
    ptr == C_NULL && error("Wrapping NULL-pointer!")
    result = Base.unsafe_wrap(Array, ptr, shape; own=false)

    if own
        finalizer(r -> API.lammps_free(ptr), result)
    end

    return copy ? Base.copy(result) : result
end
```
seems to work. It looks like the memory usage flucutaes a bit more this way than simply setting `own=true` for `Base.unsafe_wrap`, but the overall memory usage stays about the same

Comment (Joroks) [reply src/LAMMPS.jl:253]: running the test script without free of any sort:

grafik

running the test script and using `own=true` on `Base.unsafe_wrap`:

grafik

running the test script and registering `lammps_free` as a finalizer:

grafik

Comment (Joroks) [reply src/LAMMPS.jl:253]: this is probably better:

```julia
function _extract(ptr::Ptr{<:Real}, shape::Integer; copy=false, own=false)
    ptr == C_NULL && error("Wrapping NULL-pointer!")
    result = Base.unsafe_wrap(Array, ptr, shape; own=false)

    if own
        finalizer(API.lammps_free, result)
    end

    return copy ? Base.copy(result) : result
end
```

I was worried about what happens for resize operations, but julia just doesn't allow them for shared data so that's not a problem

Comment (vchuravy) [reply src/LAMMPS.jl:253]: That's an interesting idea. I am wondering about `reshape` IIRC there are some operations that return an array again, but yes if the Array is shared, they probably should not return a new Array...

Also 1.11 is introducing `Memory` and I think the finalizer will need to be attached to that.

Comment (Joroks) [reply src/LAMMPS.jl:253]: It appears like the finallizer is only run, when the underlying data is no longer acessible. So having a reshaped version of an array available would block the execution of the finalizer´.

Comment (Joroks) [reply src/LAMMPS.jl:253]: I've done:
```julia
finalizer(result) do r
    API.lammps_free(result)
    @async println("finializer has run!")
end
```
and then the following test script:

```julia
lmp = LMP()

command(lmp, """
   atom_modify map yes
   region cell block 0 3 0 3 0 3
   create_box 1 cell
   lattice sc 1
   create_atoms 1 region cell
   mass 1 1

   variable var atom id
""")

var1 = extract_variable(lmp, "var", VAR_ATOM)
var2 = reshape(var1, 1, 27)

var1 = nothing
GC.gc() # finallizer is not run

var2 = nothing
GC.gc() # finallizer is run
```

Comment (Joroks) [reply src/LAMMPS.jl:253]: I think the finalizer is the best solution. Should we merge this PR?
Then as soon as 1.11 gets released, we should revisit this. It might become necessary to force 1.11 or higher if we need to register the finalizer on the `Memory`?

Comment (vchuravy) [reply src/LAMMPS.jl:253]: So on 1.11

```
julia> var2 = reshape(var1, 1, 27)
1×27 Matrix{Float64}:
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  9.0  10.0  11.0  12.0  13.0  14.0  …  17.0  18.0  19.0  20.0  21.0  22.0  23.0  24.0  25.0  26.0  27.0

julia> var1 = nothing

julia> GC.gc() # finallizer is run

julia> var2
1×27 Matrix{Float64}:
 1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  9.0  10.0  11.0  12.0  13.0  14.0  …  18.0  19.0  20.0  21.0  22.0  23.0  24.0  25.0  26.0  1.265e-321
```

Note the data flipping.

Comment (vchuravy) [reply src/LAMMPS.jl:253]: I opened an upstream issue, but I don't expect this behavior to material change before 1.11 gets released

https://github.com/JuliaLang/julia/issues/54985

So if we want to use this finalizer version, we would need to use `@static if VERSION >= v1.11-dev` and attach the finalizer to `result.ref.mem`.

Comment (Joroks) [reply src/LAMMPS.jl:253]: this now puts a lot of stress on the GC  (in 1.11)

grafik

after running my test loop (with 10^8 iterations) i still need manually trigger GC twice for the memory to be freed completely.

grafik

this doesn't happen if I attach the finalizer to the array instead of the memory

File: src/LAMMPS.jl [hunk 14]
```diff
@@ -177,240 +245,353 @@ function get_natoms(lmp::LMP)
...
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    result = Base.unsafe_wrap(Array, ptr, shape; own=false)
 
-    ptr = API.lammps_extract_global(lmp, name)
-    ptr = reinterpret(type, ptr)
-
-    if ptr !== C_NULL
-        if dtype == API.LAMMPS_STRING
-            return Base.unsafe_string(ptr)
+    if own
```
Comment (vchuravy) [inline src/LAMMPS.jl:257]: if `copy && own` we should immediately free the data instead of attaching a finalizer.

**Review by vchuravy — APPROVED** [review]
